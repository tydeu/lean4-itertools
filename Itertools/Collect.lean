/-
Copyright (c) 2021 Mac Malone. All rights reserved.
Released under the MIT license as described in the file LICENSE.
Authors: Mac Malone
-/

namespace Itertools

/-- Produce a `List` from a monadic iterable in reverse order. -/
@[inline] def revListM [ForIn m ρ α] [Monad m] (iter : ρ) : m (List α) := do
  let mut lst := []
  for a in iter do
    lst := a :: lst
  return lst

/-- Produce a `List` from a pure iterable in reverse order. -/
@[inline] def revList [ForIn Id ρ α] (iter : ρ) : List α :=
  Id.run <| revListM iter

/-- Produce a `List` from a monadic iterable. -/
@[inline] def listM [ForIn m ρ α] [Monad m] (iter : ρ) : m (List α) :=
  (·.reverse) <$> revListM iter

/-- Produce a `List` from a pure iterable. -/
@[inline] def list [ForIn Id ρ α] (iter : ρ) : List α :=
  revList iter |>.reverse

instance [ForIn Id ρ α] : Coe ρ (List α) := ⟨list⟩

/-- Produce an `Array` from a monadic iterable. -/
@[inline] def arrayM [ForIn m ρ α] [Monad m] (iter : ρ) : m (Array α) := do
  let mut arr := #[]
  for a in iter do
    arr := arr.push a
  return arr

/-- Produce an `Array` from a pure iterable. -/
@[inline] def array [ForIn Id ρ α] (iter : ρ) : Array α :=
  Id.run <| arrayM iter

instance [ForIn Id ρ α] : Coe ρ (Array α) := ⟨array⟩
