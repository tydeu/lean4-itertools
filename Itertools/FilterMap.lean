/-
Copyright (c) 2021 Mac Malone. All rights reserved.
Released under the MIT license as described in the file LICENSE.
Authors: Mac Malone
-/

namespace Itertools

--------------------------------------------------------------------------------
-- # Filter
--------------------------------------------------------------------------------

/-- A monadic iterable transformer that performs a filter. -/
def MFilter {m : Type → Type u} {α : Type}
  (predM : α → m Bool) (ρ : Type v) := ρ

namespace MFilter
variable {m : Type → Type v} {α : Type} {predM : α → m Bool}

@[inline] def iter (self : MFilter predM ρ) : ρ := self

@[inline] protected def forIn [ForIn m ρ α] [Monad m]
(self : MFilter predM ρ) (init : β) (f : α → β → m (ForInStep β)) : m β :=
  forIn self.iter init fun a b => do
    if (← predM a) then f a b else return ForInStep.yield b

instance [ForIn m ρ α] : ForIn m (MFilter predM ρ) α := ⟨MFilter.forIn⟩
end MFilter

/-- A pure iterable transformer that performs a filter. -/
abbrev Filter {α : Type} (pred : α → Bool) (ρ : Type u) :=
  MFilter (m := Id) pred ρ

@[inline] nonrec def Filter.pred {pred : α → Bool} (_ : Filter pred ρ) := pred

/-- Filter values from `iter` that do not satisfy `predM`. -/
@[inline] def filterM {m : Type → Type u} (predM : α → m Bool) (iter : ρ) :
  MFilter predM ρ := iter

/-- Filter values from `iter` that do not satisfy `pred`. -/
@[inline] def filter (pred : α → Bool) (iter : ρ) : Filter pred ρ := iter

--------------------------------------------------------------------------------
-- # Map
--------------------------------------------------------------------------------

/-- A monadic iterable transformer that performs a map. -/
def MMap {m : Type u → Type u'} {α : Type v} {α' : Type u}
  (fnM : α → m α') (ρ : Type w) := ρ

namespace MMap
variable {m : Type u → Type v} {α : Type v} {α' : Type u} {fnM : α → m α'}

@[inline] def iter (self : MMap fnM ρ) : ρ := self

@[inline] protected def forIn [ForIn m ρ α] [Monad m]
(self : MMap fnM ρ) (init : β) (f : α' → β → m (ForInStep β)) : m β :=
  forIn self.iter init fun a b => do f (← fnM a) b

instance [ForIn m ρ α] : ForIn m (MMap fnM ρ) α' := ⟨MMap.forIn⟩
end MMap

/-- A pure iterable transformer that performs a map. -/
abbrev Map {α : Type u} {α' : Type v} (fn : α → α') (ρ : Type w) :=
  MMap (m := Id) fn ρ

@[inline] nonrec def Map.fn {fn : α → α'} (_ : Map fn ρ) := fn

/-- Apply `fnM` to each value from `iter` to produce a new iterable. -/
@[inline] def mapM {m : Type → Type u} (fnM : α → m α') (iter : ρ) :
  MMap fnM ρ := iter

/-- Apply `fn` to each value from `iter` to produce a new iterable. -/
@[inline] def map (fn : α → α') (iter : ρ) : Map fn ρ := iter
