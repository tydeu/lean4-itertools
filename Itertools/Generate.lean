/-
Copyright (c) 2021 Mac Malone. All rights reserved.
Released under the MIT license as described in the file LICENSE.
Authors: Mac Malone
-/

namespace Itertools

--------------------------------------------------------------------------------
-- # Partial Generator
--------------------------------------------------------------------------------

/-- A monadic generator that stops if/when wither `initM` or `stepM` return `none`. -/
def MGenerateSome {m : Type u → Type v} {α : Type u}
  (initM : m (Option α)) (stepM : α → m (Option α)) : Type := PUnit

namespace MGenerateSome
variable {m : Type u → Type v} {α : Type u}
variable {initM : m (Option α)} {stepM : α → m (Option α)}

@[inline] nonrec def initM (self : MGenerateSome initM stepM) := initM
@[inline] nonrec def stepM (self : MGenerateSome initM stepM) := stepM

@[inline] protected partial def forIn [Monad m]
(self : MGenerateSome initM stepM) (init : β) (f : α → β → m (ForInStep β)) : m β := do
  let rec @[specialize] loop a b := do
    match (← f a b) with
    | ForInStep.yield b =>
      match (← stepM a) with
      | some a => loop a b
      | none => pure b
    | ForInStep.done b => pure b
  match (← initM) with
  | some a => loop a init
  | none => pure init

instance : ForIn m (MGenerateSome initM stepM) α := ⟨MGenerateSome.forIn⟩
end MGenerateSome

/-- A pure generator that stops if/when wither `init` or `step` returns `none`. -/
abbrev GenerateSome {α : Type u} (init : Option α) (step : α → Option α)  :=
  MGenerateSome (m := Id) init step

namespace GenerateSome
variable {α : Type u} {init : Option α} {step : α → Option α}
@[inline] nonrec def init (self : GenerateSome init step) := init
@[inline] nonrec def step (self : GenerateSome init step) := step
end GenerateSome

/--
Generate a iterable of values within the monad `m`,
starting with `init` and continuing with `step` until either returns `none`.
-/
@[inline] def generateSomeM {m : Type u → Type v}
  (initM : m (Option α)) (stepM : α → m (Option α)) :
  MGenerateSome initM stepM := ⟨⟩

/--
Generate a iterable of values,
starting with `init` and continuing with `step` until either returns `none`.
-/
@[inline] def generateSome (init : Option α) (step : α → Option α) :
  GenerateSome init step := ⟨⟩

--------------------------------------------------------------------------------
-- # Infinite Generator
--------------------------------------------------------------------------------

/- A monadic generator, starts with `initM` and progresses with `stepM` forever. -/
def MGenerate {m : Type u → Type v} {α : Type u}
  (initM : m α) (stepM : α → m α) : Type := PUnit

namespace MGenerate
variable {m : Type u → Type v} {α : Type u}
variable {initM : m α} {stepM : α → m α}

@[inline] nonrec def initM (self : MGenerate initM stepM) := initM
@[inline] nonrec def stepM (self : MGenerate initM stepM) := stepM

@[inline] protected partial def forIn [Monad m]
(self : MGenerate initM stepM) (init : β) (f : α → β → m (ForInStep β)) : m β := do
  let rec @[specialize] loop a b := do
    match (← f a b) with
    | ForInStep.yield b => loop (← stepM a) b
    | ForInStep.done b => pure b
  loop (← initM) init

instance : ForIn m (MGenerate initM stepM) α := ⟨MGenerate.forIn⟩
end MGenerate

/- A pure generator, starts with `init` and progresses with `step` forever. -/
abbrev Generate {α : Type u} (init : α) (step : α → α)  :=
  MGenerate (m := Id) init step

namespace Generate
variable {α : Type u} {init : α} {step : α → α}
@[inline] nonrec def init (self : Generate init step) := init
@[inline] nonrec def step (self : Generate init step) := step
end Generate

/--
Generate a iterable of values within the monad `m`,
starting with `initM` and continuing infinitely with `stepM`.
-/
@[inline] def generateM {m : Type u → Type v}
  (initM : m α) (stepM : α → m α) :
  MGenerate initM stepM := ⟨⟩

/--
Generate a iterable of values,
starting with `init` and continuing infinitely with `step`.
-/
@[inline] def generate (init : α) (step : α → α) :
  Generate init step := ⟨⟩


--------------------------------------------------------------------------------
-- # Generator Transformer
--------------------------------------------------------------------------------

/-- A monadic iterable transformer, that generates values alongside its iterable. -/
def MGenerateEach {σ : Type u} {m : Type u → Type v}
  (initM : m σ) (stepM : σ → m σ) (ρ : Type w) := ρ

namespace MGenerateEach
variable {σ : Type u} {m : Type u → Type v} {ρ : Type w}
variable {initM : m σ} {stepM : σ → m σ}

@[inline] def iter (self : MGenerateEach initM stepM ρ) : ρ := self
@[inline] nonrec def initM (self : MGenerateEach initM stepM ρ) := initM
@[inline] nonrec def stepM (self : MGenerateEach initM stepM ρ) := stepM

@[inline] protected def forIn [ForIn m ρ α] [Monad m]
(self : MGenerateEach initM stepM ρ) (init : β) (f : (σ × α) → β → m (ForInStep β)) : m β := do
  Prod.snd <$> forIn self.iter (← initM, init) fun a (i, b) => do
    match (← f (i, a) b) with
    | ForInStep.yield b => ForInStep.yield (← stepM i, b)
    | ForInStep.done b => ForInStep.done (i, b)

instance [ForIn m ρ α] : ForIn m (MGenerateEach initM stepM ρ) (σ × α) := ⟨MGenerateEach.forIn⟩
end MGenerateEach

/-- A pure iterable transformer, that generates values alongside its iterable. -/
abbrev GenerateEach {σ : Type u} (init : σ) (step : σ → σ) (ρ : Type v) :=
  MGenerateEach (m := Id) init step ρ

namespace GenerateEach
variable {σ : Type u} {init : σ} {step : σ → σ} {ρ : Type v}
@[inline] nonrec def init (self : GenerateEach init step ρ) := init
@[inline] nonrec def step (self : GenerateEach init step ρ) := step
end GenerateEach

/-- Generate values alongside `iter`, producing a iterable of the product. -/
def generateEachM {m : Type u → Type v} (initM : m σ) (stepM : σ → m σ) (iter : ρ) :
  MGenerateEach initM stepM ρ := iter

/-- Generate values alongside `iter`, producing a iterable of the product. -/
def generateEach (init : σ) (step : σ → σ) (iter : ρ) :
  GenerateEach init step ρ := iter

/-- Attach an index starting at `start` and stepping by `step` to `iter`. -/
def enumerate (iter : ρ) (start := 0) (step := 1) :=
  iter |> generateEach start fun n => n + step
