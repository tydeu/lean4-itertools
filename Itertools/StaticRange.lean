/-
Copyright (c) 2021 Mac Malone. All rights reserved.
Released under the MIT license as described in the file LICENSE.
Authors: Mac Malone
-/

namespace Itertools

/-- A type-level `Std.Range`. -/
def StaticRange {α : Type u} (start stop step : α) : Type := PUnit

namespace StaticRange
variable {α : Type u} {start stop step : α}

variable [LE α] [(a b : α) → Decidable (a ≤ b)] [Add α]

@[inline] protected partial def forIn [Monad m]
(self : StaticRange start stop step) (init : β) (f : α → β → m (ForInStep β)) : m β := do
  let rec @[specialize] loop i b := do
    if i ≥ stop then
      pure b
    else
      match (← f i b) with
      | ForInStep.done b => pure b
      | ForInStep.yield b => loop (i + step) b
  loop start init

instance : ForIn m (StaticRange start stop step) α := ⟨StaticRange.forIn⟩
end StaticRange

/-- Generate an iterable static range of numeric values. -/
@[inline] def range [OfNat α 0] [OfNat α 1] (stop : α) (start : α := 0) (step : α := 1) :
  StaticRange start stop step := ⟨⟩
