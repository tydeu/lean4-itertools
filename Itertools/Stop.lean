/-
Copyright (c) 2021 Mac Malone. All rights reserved.
Released under the MIT license as described in the file LICENSE.
Authors: Mac Malone
-/

namespace Itertools

/-- A monadic iterable transformer that iterates until `predM` is satisfied. -/
def MStopIf {m : Type → Type u} {α : Type}
  (predM : α → m Bool) (ρ : Type v) := ρ

namespace MStopIf
variable {m : Type → Type v} {α : Type} {predM : α → m Bool}

@[inline] def iter (self : MStopIf predM ρ) : ρ := self
@[inline] protected def forIn [ForIn m ρ α] [Monad m]
(self : MStopIf predM ρ) (init : β) (f : α → β → m (ForInStep β)) : m β :=
  forIn self.iter init fun a b => do
    if (← predM a) then return ForInStep.done b else f a b

instance [ForIn m ρ α] : ForIn m (MStopIf predM ρ) α := ⟨MStopIf.forIn⟩
end MStopIf

/-- A monadic iterable transformer that iterates until `pred` is satisfied. -/
abbrev StopIf {α : Type} (pred : α → Bool) (ρ : Type u) :=
  MStopIf (m := Id) pred ρ

/-- Iterate until `predM` is satisfied. -/
def stopIfM {m : Type → Type u} (predM : α → m Bool) (iter : ρ) :
  MStopIf predM ρ := iter

/-- Iterate until `pred` is satisfied. -/
def stopIf (pred : α → Bool) (iter : ρ) : StopIf pred ρ := iter
