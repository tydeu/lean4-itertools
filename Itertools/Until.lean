/-
Copyright (c) 2021 Mac Malone. All rights reserved.
Released under the MIT license as described in the file LICENSE.
Authors: Mac Malone
-/

namespace Itertools

/-- A monadic iterable transformer that iterates until `predM` is satisfied. -/
def MUntil {m : Type → Type u} {α : Type}
  (predM : α → m Bool) (ρ : Type v) := ρ

namespace MUntil
variable {m : Type → Type v} {α : Type} {predM : α → m Bool}

@[inline] def iter (self : MUntil predM ρ) : ρ := self
@[inline] nonrec def predM (self : MUntil predM ρ) := predM

@[inline] protected def forIn [ForIn m ρ α] [Monad m]
(self : MUntil predM ρ) (init : β) (f : α → β → m (ForInStep β)) : m β :=
  forIn self.iter init fun a b => do
    if (← predM a) then ForInStep.done b else f a b

instance [ForIn m ρ α] : ForIn m (MUntil predM ρ) α := ⟨MUntil.forIn⟩
end MUntil

/-- A monadic iterable transformer that iterates until `pred` is satisfied. -/
abbrev Until {α : Type} (pred : α → Bool) (ρ : Type u) :=
  MUntil (m := Id) pred ρ

/-- Iterate until `predM` is satisfied. -/
def untilM {m : Type → Type u} (predM : α → m Bool) (iter : ρ) :
  MUntil predM ρ := iter

/-- Iterate until `pred` is satisfied. -/
def until (pred : α → Bool) (iter : ρ) : Until pred ρ := iter
