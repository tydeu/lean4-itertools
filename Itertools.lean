/-
Copyright (c) 2021 Mac Malone. All rights reserved.
Released under the MIT license as described in the file LICENSE.
Authors: Mac Malone
-/
import Itertools.Generate
import Itertools.StaticRange
import Itertools.Until
import Itertools.FilterMap
import Itertools.Collect
