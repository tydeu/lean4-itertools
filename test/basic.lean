/-
Copyright (c) 2021 Mac Malone. All rights reserved.
Released under the MIT license as described in the file LICENSE.
Authors: Mac Malone
-/

import Itertools

open Itertools

set_option trace.compiler.ir.result true

def foo :=
  generate 0 (· + 1)
  |> until (· == 10)
  |> filter (· % 2 == 0)
  |> map (· * 2)
  |> array

def test1 :=
  array <| filter (fun i => i % 2 == 0) [0:5]

def test2 :=
  array <| filter (fun (i, (_ : Nat)) => i % 2 == 0) <| enumerate [0:5]

def test3 :=
  array <| enumerate [0:5]

def test4 :=
  generate 0 (· + 1)
  |> until (· == 10)
  |> array

def testInf :=
  generate 0 (· + 1)
  |> array

set_option trace.compiler.ir.result false

#eval foo
#eval test1
#eval test2
#eval test3
#eval test4
