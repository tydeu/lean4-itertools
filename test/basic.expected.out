
[result]
def Itertools.MGenerate.forIn.loop._at.foo._spec_1 (x_1 : obj) (x_2 : obj) : obj :=
  let x_3 : obj := 10;
  let x_4 : u8 := Nat.beq x_1 x_3;
  case x_4 : u8 of
  Bool.false →
    let x_5 : obj := 2;
    let x_6 : obj := Nat.mod x_1 x_5;
    let x_7 : obj := 0;
    let x_8 : u8 := Nat.beq x_6 x_7;
    dec x_6;
    case x_8 : u8 of
    Bool.false →
      let x_9 : obj := 1;
      let x_10 : obj := Nat.add x_1 x_9;
      dec x_1;
      let x_11 : obj := Itertools.MGenerate.forIn.loop._at.foo._spec_1 x_10 x_2;
      ret x_11
    Bool.true →
      let x_12 : obj := Nat.mul x_1 x_5;
      let x_13 : obj := Array.push ◾ x_2 x_12;
      let x_14 : obj := 1;
      let x_15 : obj := Nat.add x_1 x_14;
      dec x_1;
      let x_16 : obj := Itertools.MGenerate.forIn.loop._at.foo._spec_1 x_15 x_13;
      ret x_16
  Bool.true →
    dec x_1;
    ret x_2
def foo._closed_1 : obj :=
  let x_1 : obj := 0;
  let x_2 : obj := Array.mkEmpty ◾ x_1;
  ret x_2
def foo : obj :=
  let x_1 : obj := 0;
  let x_2 : obj := foo._closed_1;
  let x_3 : obj := Itertools.MGenerate.forIn.loop._at.foo._spec_1 x_1 x_2;
  ret x_3
[result]
def Itertools.StaticRange.forIn.loop._at.foo'._spec_1 (x_1 : @& obj) (x_2 : @& obj) (x_3 : obj) (x_4 : obj) : obj :=
  let x_5 : u8 := Nat.decLe x_1 x_3;
  case x_5 : u8 of
  Bool.false →
    let x_6 : obj := 2;
    let x_7 : obj := Nat.mod x_3 x_6;
    let x_8 : obj := 0;
    let x_9 : u8 := Nat.beq x_7 x_8;
    dec x_7;
    case x_9 : u8 of
    Bool.false →
      let x_10 : obj := Nat.add x_3 x_2;
      dec x_3;
      let x_11 : obj := Itertools.StaticRange.forIn.loop._at.foo'._spec_1 x_1 x_2 x_10 x_4;
      ret x_11
    Bool.true →
      let x_12 : obj := Nat.mul x_3 x_6;
      let x_13 : obj := Array.push ◾ x_4 x_12;
      let x_14 : obj := Nat.add x_3 x_2;
      dec x_3;
      let x_15 : obj := Itertools.StaticRange.forIn.loop._at.foo'._spec_1 x_1 x_2 x_14 x_13;
      ret x_15
  Bool.true →
    dec x_3;
    ret x_4
def foo' : obj :=
  let x_1 : obj := 10;
  let x_2 : obj := 1;
  let x_3 : obj := 0;
  let x_4 : obj := foo._closed_1;
  let x_5 : obj := Itertools.StaticRange.forIn.loop._at.foo'._spec_1 x_1 x_2 x_3 x_4;
  ret x_5
def Itertools.StaticRange.forIn.loop._at.foo'._spec_1._boxed (x_1 : obj) (x_2 : obj) (x_3 : obj) (x_4 : obj) : obj :=
  let x_5 : obj := Itertools.StaticRange.forIn.loop._at.foo'._spec_1 x_1 x_2 x_3 x_4;
  dec x_2;
  dec x_1;
  ret x_5
[result]
def Std.Range.forIn.loop._at.test1._spec_1 (x_1 : obj) (x_2 : obj) (x_3 : @& obj) (x_4 : @& obj) (x_5 : obj) : obj :=
  let x_6 : u8 := Nat.decLe x_3 x_2;
  case x_6 : u8 of
  Bool.false →
    let x_7 : obj := 0;
    let x_8 : u8 := Nat.decEq x_1 x_7;
    case x_8 : u8 of
    Bool.false →
      let x_9 : obj := 1;
      let x_10 : obj := Nat.sub x_1 x_9;
      dec x_1;
      let x_11 : obj := 2;
      let x_12 : obj := Nat.mod x_2 x_11;
      let x_13 : u8 := Nat.beq x_12 x_7;
      dec x_12;
      case x_13 : u8 of
      Bool.false →
        let x_14 : obj := Nat.add x_2 x_4;
        dec x_2;
        let x_15 : obj := Std.Range.forIn.loop._at.test1._spec_1 x_10 x_14 x_3 x_4 x_5;
        ret x_15
      Bool.true →
        inc x_2;
        let x_16 : obj := Array.push ◾ x_5 x_2;
        let x_17 : obj := Nat.add x_2 x_4;
        dec x_2;
        let x_18 : obj := Std.Range.forIn.loop._at.test1._spec_1 x_10 x_17 x_3 x_4 x_16;
        ret x_18
    Bool.true →
      dec x_2;
      dec x_1;
      ret x_5
  Bool.true →
    dec x_2;
    dec x_1;
    ret x_5
def test1 : obj :=
  let x_1 : obj := 5;
  let x_2 : obj := 0;
  let x_3 : obj := 1;
  let x_4 : obj := foo._closed_1;
  let x_5 : obj := Std.Range.forIn.loop._at.test1._spec_1 x_1 x_2 x_1 x_3 x_4;
  ret x_5
def Std.Range.forIn.loop._at.test1._spec_1._boxed (x_1 : obj) (x_2 : obj) (x_3 : obj) (x_4 : obj) (x_5 : obj) : obj :=
  let x_6 : obj := Std.Range.forIn.loop._at.test1._spec_1 x_1 x_2 x_3 x_4 x_5;
  dec x_4;
  dec x_3;
  ret x_6
[result]
def Std.Range.forIn.loop._at.test2._spec_1 (x_1 : obj) (x_2 : obj) (x_3 : @& obj) (x_4 : @& obj) (x_5 : obj) : obj :=
  let x_6 : u8 := Nat.decLe x_3 x_2;
  case x_6 : u8 of
  Bool.false →
    let x_7 : obj := 0;
    let x_8 : u8 := Nat.decEq x_1 x_7;
    case x_8 : u8 of
    Bool.false →
      let x_9 : obj := 1;
      let x_10 : obj := Nat.sub x_1 x_9;
      dec x_1;
      let x_11 : u8 := isShared x_5;
      case x_11 : u8 of
      Bool.false →
        let x_12 : obj := proj[0] x_5;
        let x_13 : obj := proj[1] x_5;
        inc x_2;
        inc x_12;
        set x_5[1] := x_2;
        let x_14 : obj := 2;
        let x_15 : obj := Nat.mod x_12 x_14;
        let x_16 : u8 := Nat.beq x_15 x_7;
        dec x_15;
        case x_16 : u8 of
        Bool.false →
          dec x_5;
          let x_17 : obj := Nat.add x_12 x_9;
          dec x_12;
          let x_18 : obj := ctor_0[Prod.mk] x_17 x_13;
          let x_19 : obj := Nat.add x_2 x_4;
          dec x_2;
          let x_20 : obj := Std.Range.forIn.loop._at.test2._spec_1 x_10 x_19 x_3 x_4 x_18;
          ret x_20
        Bool.true →
          let x_21 : obj := Array.push ◾ x_13 x_5;
          let x_22 : obj := Nat.add x_12 x_9;
          dec x_12;
          let x_23 : obj := ctor_0[Prod.mk] x_22 x_21;
          let x_24 : obj := Nat.add x_2 x_4;
          dec x_2;
          let x_25 : obj := Std.Range.forIn.loop._at.test2._spec_1 x_10 x_24 x_3 x_4 x_23;
          ret x_25
      Bool.true →
        let x_26 : obj := proj[0] x_5;
        let x_27 : obj := proj[1] x_5;
        inc x_27;
        inc x_26;
        dec x_5;
        inc x_2;
        inc x_26;
        let x_28 : obj := ctor_0[Prod.mk] x_26 x_2;
        let x_29 : obj := 2;
        let x_30 : obj := Nat.mod x_26 x_29;
        let x_31 : u8 := Nat.beq x_30 x_7;
        dec x_30;
        case x_31 : u8 of
        Bool.false →
          dec x_28;
          let x_32 : obj := Nat.add x_26 x_9;
          dec x_26;
          let x_33 : obj := ctor_0[Prod.mk] x_32 x_27;
          let x_34 : obj := Nat.add x_2 x_4;
          dec x_2;
          let x_35 : obj := Std.Range.forIn.loop._at.test2._spec_1 x_10 x_34 x_3 x_4 x_33;
          ret x_35
        Bool.true →
          let x_36 : obj := Array.push ◾ x_27 x_28;
          let x_37 : obj := Nat.add x_26 x_9;
          dec x_26;
          let x_38 : obj := ctor_0[Prod.mk] x_37 x_36;
          let x_39 : obj := Nat.add x_2 x_4;
          dec x_2;
          let x_40 : obj := Std.Range.forIn.loop._at.test2._spec_1 x_10 x_39 x_3 x_4 x_38;
          ret x_40
    Bool.true →
      dec x_2;
      dec x_1;
      ret x_5
  Bool.true →
    dec x_2;
    dec x_1;
    ret x_5
def test2._closed_1 : obj :=
  let x_1 : obj := 0;
  let x_2 : obj := foo._closed_1;
  let x_3 : obj := ctor_0[Prod.mk] x_1 x_2;
  ret x_3
def test2 : obj :=
  let x_1 : obj := 5;
  let x_2 : obj := 0;
  let x_3 : obj := 1;
  let x_4 : obj := test2._closed_1;
  let x_5 : obj := Std.Range.forIn.loop._at.test2._spec_1 x_1 x_2 x_1 x_3 x_4;
  let x_6 : obj := proj[1] x_5;
  inc x_6;
  dec x_5;
  ret x_6
def Std.Range.forIn.loop._at.test2._spec_1._boxed (x_1 : obj) (x_2 : obj) (x_3 : obj) (x_4 : obj) (x_5 : obj) : obj :=
  let x_6 : obj := Std.Range.forIn.loop._at.test2._spec_1 x_1 x_2 x_3 x_4 x_5;
  dec x_4;
  dec x_3;
  ret x_6
[result]
def Std.Range.forIn.loop._at.test3._spec_1 (x_1 : obj) (x_2 : obj) (x_3 : @& obj) (x_4 : @& obj) (x_5 : obj) : obj :=
  let x_6 : u8 := Nat.decLe x_3 x_2;
  case x_6 : u8 of
  Bool.false →
    let x_7 : obj := 0;
    let x_8 : u8 := Nat.decEq x_1 x_7;
    case x_8 : u8 of
    Bool.false →
      let x_9 : obj := 1;
      let x_10 : obj := Nat.sub x_1 x_9;
      dec x_1;
      let x_11 : u8 := isShared x_5;
      case x_11 : u8 of
      Bool.false →
        let x_12 : obj := proj[0] x_5;
        let x_13 : obj := proj[1] x_5;
        inc x_2;
        inc x_12;
        set x_5[1] := x_2;
        let x_14 : obj := Array.push ◾ x_13 x_5;
        let x_15 : obj := Nat.add x_12 x_9;
        dec x_12;
        let x_16 : obj := ctor_0[Prod.mk] x_15 x_14;
        let x_17 : obj := Nat.add x_2 x_4;
        dec x_2;
        let x_18 : obj := Std.Range.forIn.loop._at.test3._spec_1 x_10 x_17 x_3 x_4 x_16;
        ret x_18
      Bool.true →
        let x_19 : obj := proj[0] x_5;
        let x_20 : obj := proj[1] x_5;
        inc x_20;
        inc x_19;
        dec x_5;
        inc x_2;
        inc x_19;
        let x_21 : obj := ctor_0[Prod.mk] x_19 x_2;
        let x_22 : obj := Array.push ◾ x_20 x_21;
        let x_23 : obj := Nat.add x_19 x_9;
        dec x_19;
        let x_24 : obj := ctor_0[Prod.mk] x_23 x_22;
        let x_25 : obj := Nat.add x_2 x_4;
        dec x_2;
        let x_26 : obj := Std.Range.forIn.loop._at.test3._spec_1 x_10 x_25 x_3 x_4 x_24;
        ret x_26
    Bool.true →
      dec x_2;
      dec x_1;
      ret x_5
  Bool.true →
    dec x_2;
    dec x_1;
    ret x_5
def test3 : obj :=
  let x_1 : obj := 5;
  let x_2 : obj := 0;
  let x_3 : obj := 1;
  let x_4 : obj := test2._closed_1;
  let x_5 : obj := Std.Range.forIn.loop._at.test3._spec_1 x_1 x_2 x_1 x_3 x_4;
  let x_6 : obj := proj[1] x_5;
  inc x_6;
  dec x_5;
  ret x_6
def Std.Range.forIn.loop._at.test3._spec_1._boxed (x_1 : obj) (x_2 : obj) (x_3 : obj) (x_4 : obj) (x_5 : obj) : obj :=
  let x_6 : obj := Std.Range.forIn.loop._at.test3._spec_1 x_1 x_2 x_3 x_4 x_5;
  dec x_4;
  dec x_3;
  ret x_6
[result]
def Itertools.MGenerate.forIn.loop._at.test4._spec_1 (x_1 : obj) (x_2 : obj) : obj :=
  let x_3 : obj := 10;
  let x_4 : u8 := Nat.beq x_1 x_3;
  case x_4 : u8 of
  Bool.false →
    inc x_1;
    let x_5 : obj := Array.push ◾ x_2 x_1;
    let x_6 : obj := 1;
    let x_7 : obj := Nat.add x_1 x_6;
    dec x_1;
    let x_8 : obj := Itertools.MGenerate.forIn.loop._at.test4._spec_1 x_7 x_5;
    ret x_8
  Bool.true →
    dec x_1;
    ret x_2
def test4 : obj :=
  let x_1 : obj := 0;
  let x_2 : obj := foo._closed_1;
  let x_3 : obj := Itertools.MGenerate.forIn.loop._at.test4._spec_1 x_1 x_2;
  ret x_3
[result]
def Itertools.MGenerate.forIn.loop._at.testInf._spec_1 (x_1 : obj) (x_2 : obj) : obj :=
  inc x_1;
  let x_3 : obj := Array.push ◾ x_2 x_1;
  let x_4 : obj := 1;
  let x_5 : obj := Nat.add x_1 x_4;
  dec x_1;
  let x_6 : obj := Itertools.MGenerate.forIn.loop._at.testInf._spec_1 x_5 x_3;
  ret x_6
def testInf : obj :=
  let x_1 : obj := 0;
  let x_2 : obj := foo._closed_1;
  let x_3 : obj := Itertools.MGenerate.forIn.loop._at.testInf._spec_1 x_1 x_2;
  ret x_3#[0, 4, 8, 12, 16]
#[0, 4, 8, 12, 16]
#[0, 2, 4]
#[(0, 0), (2, 2), (4, 4)]
#[(0, 0), (1, 1), (2, 2), (3, 3), (4, 4)]
#[0, 1, 2, 3, 4, 5, 6, 7, 8, 9]
